### DISCUSSION TITLE
Load Tests on Smallstep Not Making Sense

### DISCUSSION THREAD

**Discussion by TheSecMaven** [smallstep/certificates#417] [Q&A]
Running load tests using the CLI.

1. check health
2. login, get a 1 time token
3. use one time token to get a cert
4. revoke that cert
5. repeat

We get to around 7 TPS and just maxes out. CPU is not floored, neither is Memory, Same for the Mysql DB, just shows some CPU but nothing extravagant.

What are some thoughts on things we may need to tune? Best we can tell, nothing OS level is hitting an issue.

**Comment by maraino** [comment]
The use of `step` for a load test doesn't make any sense. You should either pre-create hundreds/thousands of tokens (with a custom validity) or use the k8ssa provisioner so you can use the same token multiple times. So instead of doing step 2 and 3, you need to simulate a command like:
```sh
step ca certificate --k8ssa-token-path k8ssa.token internal.smallstep.com internal.crt internal.key
```

But of course not using `step`. I would suggest [vegeta](https://github.com/tsenart/vegeta)

**Reply by TheSecMaven** [reply]
hmm. You are saying its not a valid load test to be testing the creation of one time tokens and then using those tokens to get signed certs? We thought it would be a valid test of the common operations people will be doing. After all, the whole point of a load test is to do what people will do with the tool, but at scale and much faster :) 

With that said, and either way, we still only got roughly 6.5 TPS, and it doesn't make much sense. Is there another thing that could be hitting a limit, that comes to mind? I don't know enough about the inner workings to make that call, or maybe we are looking at the wrong thing?

**Reply by maraino** [reply]
I've just created using vegeta a simple test:

A file with the tests, let's call it `load.test`
```
GET https://ca.smallstep.com:9000/1.0/health

POST https://ca.smallstep.com:9000/1.0/sign
@sign.json
```

The body of the `sign` method (`sign.json`) with a token for a k8ssa provisioner and a simple csr:
```
{
    "ott": "eyJhbGciOiJFUzI1NiIsInR5cCI6IkpXVCJ9.eyJrdWJlcm5ldGVzLmlvL3NlcnZpY2VhY2NvdW50L3NlcnZpY2UtYWNjb3VudC5uYW1lIjoic21hbGxzdGVwIiwiaXNzIjoia3ViZXJuZXRlcy9zZXJ2aWNlYWNjb3VudCIsInN1YiI6InNtYWxsc3RlcCIsImlhdCI6MTU4ODk4NDA1OX0.2RFFo6Zdc-7mDVORg1Io6v2T6KKv_DCHCgiizCNx1n4ECGAvpqiv6XRpkku8DIAYNY2kAVad8WUehuyeYYMYAA",
    "csr": "-----BEGIN CERTIFICATE REQUEST-----\nMIHOMHYCAQAwFDESMBAGA1UEAxMJc21hbGxzdGVwMFkwEwYHKoZIzj0CAQYIKoZI\nzj0DAQcDQgAEKHWA08cjq5bEAZ4UPHAgKt/h73P39De2LkJ4hV+rJAbYYaKl6lNN\n08y+ExjGwpd4ns8Fwl/KDwvWngg2RoUteqAAMAoGCCqGSM49BAMCA0gAMEUCIBXX\nxHQkmf4nuiWGQVUe1VcfGFGeBaDEevU4EsC6Zq9PAiEAzNEfm7cmMirf4x8q1c6p\navYfz9zKg/+5v/aTjyCwXe8=\n-----END CERTIFICATE REQUEST-----\n"
}
```

Put everything toguether and run at 50rps:
```sh
$ cat load.test |  vegeta attack --root-certs $(step path)/certs/root_ca.crt --rate 50 --duration 30s | tee results.bin | vegeta report
Requests      [total, rate, throughput]         1500, 50.03, 50.03
Duration      [total, attack, wait]             29.983s, 29.981s, 2.394ms
Latencies     [min, mean, 50, 90, 95, 99, max]  362.35µs, 1.392ms, 1.495ms, 2.514ms, 2.774ms, 3.708ms, 13.341ms
Bytes In      [total, mean]                     2195244, 1463.50
Bytes Out     [total, mean]                     523500, 349.00
Success       [ratio]                           100.00%
Status Codes  [code:count]                      200:750  201:750
Error Set:
```

**Reply by maraino** [reply]
It requires a little bit more work to do, but you can actually get real data.

**Reply by TheSecMaven** [reply]
We are basically doing the same thing then, yours via the api directly, and mine using the cli.

Am I missing something? Again, the goal of the post was to evaluate what could be causing low throughout. Database connection pooling? Is there something that small step does as it’s connecting to the DB maybe? Perhaps someone has seen an issue if tuning on the MySQL dB isn’t done right?

**Reply by mmalone** [reply]
tl;dr: check your resource utilization client-side.

How many clients are you pointing at the CA simultaneously? The most resource-intensive parts of the process you're describing are going to be generating the random bits for your key pair and signing the one-time token and certificate signing request (CSR). Both of those processes happen client-side.

**Reply by TheSecMaven** [reply]
yep, agreed. What I didn't clarify on this was that the test was increasing the rate at which it attempted to do the test as time passed. What we see, is that we max out at 7tps, and then we continue to increase the number of vusers attempting to get a cert, and they don't have an impact. We continue to request more, and step simply says "give me a minute", those new users wait for their turn, then finish successfully. Step doesn't turn them away, just says "I can't get more than 7 tps, so you gotta wait". We are looking at OS level tuning right now

**Reply by maraino** [reply]
We're not limiting the rate of tokens that we produce, if you're using an OIDC provisioner, they might have that limit.

There always will be an extra cost, huge, of just running step to produce a token instead of creating a custom program that creates and streams JWK tokens, or just using a multi-use token as I suggested before.

A limit in MySQL will only affect how the tokens and certificates are stored at signing time, but not at token creation.

**Reply by mmalone** [reply]
Hrm. It's hard to say without knowing a lot more about your setup. I'm not sure what provisioner(s) and signature algorithm(s) you're using. I'm also not sure what a vuser is in your setup. But 7tps is one or two orders of magnitude lower than what I'd expect to see. Based on what you're describing and based on my experience with load testing, my best guess is still that you're hitting client-side resource limits. Make sure you're multi-threaded, using non-blocking IO, etc. on the client side.

You're looking in the right places: scarce resources are gonna be CPU, memory, network, or disk (see `iostat`). If you're using blocking IO then network _latency_ will also affect throughput (e.g., if it takes 100ms round-trip to respond to a request due to network latency then a blocking, single-threaded client will max out at <= 10 requests/sec). There is sometimes an additional scarce resource for crypto stuff: randomness. But, unless you're doing something really unusual, that should show up as CPU since we're using a CSPRNG under the hood and not relying on an external source of randomness (e.g., `/dev/random`).

**Reply by TheSecMaven** [reply]
any example api call for generating the token? little hard to poke through the code, as im not super famiilar.

**Reply by maraino** [reply]
The JWK code is generated in the client, we do 2 different calls:
* `GET /provisioners?limit=100` to show the list of provisioners
* `GET /provisioners/<jwk-provisioner-kid>/encrypted-key` to get the encrypted key.

Then `step` asks for a password and decrypts the key. You can do something similar with `step crypto jwe decrypt`:
```
curl -s https://my.ca/provisioners/<kid>/encrypted-key | jq -r .key | step crypto jwe decrypt
```

And that key is used to sign the token. If you store the key, encrypted or decrypted in `/tmp/my.key` you can create a token offline for `internal.smallstep.com` using:
```
echo '{"sans": ["internal.smallstep.com"],"sha":"422cce6c2b31213fa7c53c4ad163ad63dcf9be50fe8577c0dc3d4c90685d9770"}' | step crypto jwt sign -key /tmp/my.key --iss "my-issuer" --aud https://my.ca/1.0/sign -sub internal.smallstep.com -exp $(date -v+1M +"%s")
```

You can see the values for `--iss` and `--aud` inspecting a token:
```
step ca token internal.smallstep.com | step crypto jwt inspect --insecure
```

And the value for `"sha"` is the fingerprint of the root_ca.crt, but you can omit this if your environment is bootstrapped:
```
step certificate fingerprint $(step path)/certs/root_ca.crt
```

The token created now can be used with
```
step ca certificate --token $TOKEN internal.smallstep.com internal.crt internal.key
```

And if you want you can pre-create thousands of tokens using the above command, but adjusting the expiration (`--exp`) to something more than just 1m.
